src/login/loginCoordinator.ts

```diff
@@ -0,0 +1,286 @@
+						);
+
+						await this.persistSessionAuth(result, safeHostname, newUrl);
+
+						return result;
+					} else {
+						// User cancelled
+						return { success: false };
+					}
+				});
+
+			// Race between user clicking login and cross-window detection
+			return Promise.race([
+				dialogPromise,
+				this.waitForCrossWindowLogin(safeHostname),
+			]);
+		});
+	}
+
+	private async persistSessionAuth(
+		result: LoginResult,
+		safeHostname: string,
+		url: string,
+	): Promise<void> {
+		// Empty token is valid for mTLS
+		if (result.success && result.token !== undefined) {
+			await this.secretsManager.setSessionAuth(safeHostname, {
+				url,
+				token: result.token,
+			});
+			await this.mementoManager.addToUrlHistory(url);
+		}
+	}
+
+	/**
+	 * Same-window guard wrapper.
+	 */
+	private async executeWithGuard(
+		safeHostname: string,
+		executeFn: () => Promise<LoginResult>,
+	): Promise<LoginResult> {
+		const existingLogin = this.inProgressLogins.get(safeHostname);
+		if (existingLogin) {
+			return existingLogin;
+		}
+
+		const loginPromise = executeFn();
+		this.inProgressLogins.set(safeHostname, loginPromise);
+
+		try {
+			return await loginPromise;
+		} finally {
+			this.inProgressLogins.delete(safeHostname);
+		}
+	}
+
+	/**
+	 * Waits for login detected from another window.
+	 */
+	private async waitForCrossWindowLogin(
+		safeHostname: string,
+	): Promise<LoginResult> {
+		return new Promise((resolve) => {
+			const disposable = this.secretsManager.onDidChangeSessionAuth(
+				safeHostname,
+				(auth) => {
+					if (auth?.token) {
+						disposable.dispose();
+						resolve({ success: true, token: auth.token });
+					}
+				},
+			);
+		});
+	}
+
+	/**
+	 * Attempt to authenticate using token, or mTLS. If necessary, prompts
+	 * for authentication method and credentials. Returns the token and user upon
+	 * successful authentication. Null means the user aborted or authentication
+	 * failed (in which case an error notification will have been displayed).
+	 */
+	private async attemptLogin(
+		deployment: Deployment,
+		isAutoLogin: boolean,
+	): Promise<LoginResult> {
+		const needsToken = needToken(vscode.workspace.getConfiguration());
+		const client = CoderApi.create(deployment.url, "", this.logger);
+
+		let storedToken: string | undefined;
+		if (needsToken) {
+			const auth = await this.secretsManager.getSessionAuth(
+				deployment.safeHostname,
+			);
+			storedToken = auth?.token;
+			if (storedToken) {
+				client.setSessionToken(storedToken);
+			}
+		}
+
+		// Attempt authentication with current credentials (token or mTLS)
+		try {
+			const user = await client.getAuthenticatedUser();
+			// Return the token that was used (empty string for mTLS since
+			// the `vscodessh` command currently always requires a token file)
+			return { success: true, token: storedToken ?? "", user };
+		} catch (err) {
+			if (needsToken) {
+				// For token auth: silently continue to prompt for new credentials
```

Comment:
Only if the error is 401 probably?  A 500 error for example would not indicate the credentials are invalid.